<a href="https://colab.research.google.com/github/Group-10-Refactory-Capstone/Group10RefactoryCapstone/blob/main/EDA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#This is a notebook in which data exploration is going to be done

# 1. Importing libraries

In [5]:
import pandas as pd
import numpy as np
import seaborn as sns                       #package for visualisation
import matplotlib.pyplot as plt             #package for plotting graphs and images
%matplotlib inline
sns.set(color_codes=True)

# 2. Loading the train data files into dataframes

In [1]:
data_folder ='/content/drive/MyDrive/Refactory (1)/Capstone/Loan default'

## Unzipping zip files

In [2]:
!unzip '/content/drive/MyDrive/Refactory (1)/Capstone/Loan default/trainprevloans.zip'

Archive:  /content/drive/MyDrive/Refactory (1)/Capstone/Loan default/trainprevloans.zip
  inflating: trainprevloans.csv      


## Creating dataframes

In [7]:
df_train_prev_loans = pd.read_csv('/content/trainprevloans.csv')
df_train_prev_loans.head(10)

customerid  systemloanid  loannumber  \
0  8a2a81a74ce8c05d014cfb32a0da1049     301682320           2   
1  8a2a81a74ce8c05d014cfb32a0da1049     301883808           9   
2  8a2a81a74ce8c05d014cfb32a0da1049     301831714           8   
3  8a8588f35438fe12015444567666018e     301861541           5   
4  8a85890754145ace015429211b513e16     301941754           2   
5  8a858970548359cc0154883481981866     301832407           5   
6  8a858970548359cc0154883481981866     301926513           8   
7  8a8589f35451855401546b0738c42524     301912443           6   
8  8a858e095c59b91b015c5e5cea3719bc     301936857           2   
9  8a858e1158dc4d830158f7bde4f47ea7     301811291           3   

                 approveddate                creationdate  loanamount  \
0  2016-08-15 18:22:40.000000  2016-08-15 17:22:32.000000     10000.0   
1  2017-04-28 18:39:07.000000  2017-04-28 17:38:53.000000     10000.0   
2  2017-03-05 10:56:25.000000  2017-03-05 09:56:19.000000     20000.0   
3  2017-04-09 18:25:55.000000  2017-04-09 17:25:42.000000     10000.0   
4  2017-06-17 09:29:57.000000  2017-06-17 08:29:50.000000     10000.0   
5  2017-03-06 13:25:17.000000  2017-03-06 12:25:10.000000     20000.0   
6  2017-06-04 01:00:23.000000  2017-06-04 00:00:16.000000     30000.0   
7  2017-05-18 09:42:18.000000  2017-05-18 08:41:12.000000     20000.0   
8  2017-06-13 08:26:31.000000  2017-06-13 07:26:24.000000     10000.0   
9  2017-02-06 18:55:30.000000  2017-02-06 17:55:20.000000     10000.0   

   totaldue  termdays                  closeddate referredby  \
0   13000.0        30  2016-09-01 16:06:48.000000        NaN   
1   13000.0        30  2017-05-28 14:44:49.000000        NaN   
2   23800.0        30  2017-04-26 22:18:56.000000        NaN   
3   11500.0        15  2017-04-24 01:35:52.000000        NaN   
4   11500.0        15  2017-07-14 21:18:43.000000        NaN   
5   23800.0        30  2017-04-04 15:46:56.000000        NaN   
6   39000.0        60  2017-07-03 23:35:38.000000        NaN   
7   24500.0        30  2017-06-19 10:10:35.000000        NaN   
8   13000.0        30  2017-07-10 13:32:03.000000        NaN   
9   11500.0        15  2017-02-21 05:34:20.000000        NaN   

                 firstduedate             firstrepaiddate  
0  2016-09-14 00:00:00.000000  2016-09-01 15:51:43.000000  
1  2017-05-30 00:00:00.000000  2017-05-26 00:00:00.000000  
2  2017-04-04 00:00:00.000000  2017-04-26 22:03:47.000000  
3  2017-04-24 00:00:00.000000  2017-04-24 00:48:43.000000  
4  2017-07-03 00:00:00.000000  2017-07-14 21:08:35.000000  
5  2017-04-05 00:00:00.000000  2017-04-04 15:31:47.000000  
6  2017-07-04 00:00:00.000000  2017-07-03 23:25:29.000000  
7  2017-06-19 00:00:00.000000  2017-06-19 10:00:21.000000  
8  2017-07-13 00:00:00.000000  2017-07-10 13:21:53.000000  
9  2017-02-21 00:00:00.000000  2017-02-21 05:19:09.000000

In [8]:
df_train_demographics = pd.read_csv('/content/drive/MyDrive/Refactory (1)/Capstone/Loan default/traindemographics.csv')
df_train_demographics.head(10)

customerid                   birthdate  \
0  8a858e135cb22031015cbafc76964ebd  1973-10-10 00:00:00.000000   
1  8a858e275c7ea5ec015c82482d7c3996  1986-01-21 00:00:00.000000   
2  8a858e5b5bd99460015bdc95cd485634  1987-04-01 00:00:00.000000   
3  8a858efd5ca70688015cabd1f1e94b55  1991-07-19 00:00:00.000000   
4  8a858e785acd3412015acd48f4920d04  1982-11-22 00:00:00.000000   
5  8a858f045bc9690c015bca251e9a4e95  1978-12-11 00:00:00.000000   
6  8a858ea05a859123015a8892914d15b7  1990-07-21 00:00:00.000000   
7  8a858f275c451af5015c546a50ce55ed  1986-09-09 00:00:00.000000   
8  8a858f405d13c45f015d13dd93ec0c1c  1992-11-17 00:00:00.000000   
9  8a858f735c161bda015c164d86632026  1965-02-23 00:00:00.000000   

  bank_account_type  longitude_gps  latitude_gps bank_name_clients  \
0           Savings       3.319219      6.528604           GT Bank   
1           Savings       3.325598      7.119403     Sterling Bank   
2           Savings       5.746100      5.563174     Fidelity Bank   
3           Savings       3.362850      6.642485           GT Bank   
4           Savings       8.455332     11.971410           GT Bank   
5           Savings       3.330087      6.552518           GT Bank   
6           Savings       3.365935      6.564823       Access Bank   
7           Savings       4.734382      7.606868           GT Bank   
8           Savings       3.290590      6.612075           EcoBank   
9           Savings       3.362783      6.509418           EcoBank   

  bank_branch_clients employment_status_clients level_of_education_clients  
0                 NaN                       NaN                        NaN  
1                 NaN                 Permanent                        NaN  
2                 NaN                       NaN                        NaN  
3                 NaN                 Permanent                        NaN  
4                 NaN                 Permanent                        NaN  
5                 NaN                 Permanent                        NaN  
6                 NaN                 Permanent                        NaN  
7                 NaN                 Permanent                        NaN  
8                 NaN                 Permanent                        NaN  
9                 NaN                 Permanent                        NaN

In [11]:
df_train_demographics.tail(10)

customerid                   birthdate  \
4336  8a858e685c217d36015c223d219926a4  1990-02-17 00:00:00.000000   
4337  8a858fbc5cc666c8015cc9b715f420a8  1967-06-26 00:00:00.000000   
4338  8a858f735b437619015b437bf3d30bdc  1982-10-19 00:00:00.000000   
4339  8a858fc45c5f93be015c61e0e5e55e38  1986-02-16 00:00:00.000000   
4340  8a85897353d85af80153d8a368020159  1983-06-07 00:00:00.000000   
4341  8a858f155554552501555588ca2b3b40  1985-12-13 00:00:00.000000   
4342  8a858fc65cf978f4015cf97cee3a02ce  1982-07-01 00:00:00.000000   
4343  8a858f4f5b66de3a015b66fc83c61902  1989-09-26 00:00:00.000000   
4344  8aaae7a74400b28201441c8b62514150  1985-09-06 00:00:00.000000   
4345  8a85896653e2e18b0153e69c1b90265c  1975-06-05 00:00:00.000000   

     bank_account_type  longitude_gps  latitude_gps bank_name_clients  \
4336           Savings       3.343461      6.623492     Fidelity Bank   
4337           Savings       3.339679      7.126463      Stanbic IBTC   
4338           Savings       3.306874      6.669184           GT Bank   
4339           Savings       3.353483      6.651007           GT Bank   
4340             Other       7.046360      5.499944               UBA   
4341             Other       3.236753      7.030168      Stanbic IBTC   
4342           Savings       7.013750      4.875662           GT Bank   
4343           Savings       6.295530      7.092508           GT Bank   
4344           Savings       3.354206      6.539070           GT Bank   
4345           Savings       6.661014      7.472700               UBA   

     bank_branch_clients employment_status_clients level_of_education_clients  
4336                 NaN                 Permanent                        NaN  
4337                 NaN                 Permanent                        NaN  
4338                 NaN                 Permanent                        NaN  
4339                 NaN                       NaN                        NaN  
4340                 NaN             Self-Employed                        NaN  
4341                 NaN                 Permanent                   Graduate  
4342                 NaN                       NaN                        NaN  
4343                 NaN                 Permanent                        NaN  
4344         HEAD OFFICE                 Permanent                    Primary  
4345                 NaN                 Permanent                        NaN

In [9]:
df_train_perf = pd.read_csv('/content/drive/MyDrive/Refactory (1)/Capstone/Loan default/trainperf.csv')
df_train_perf.head(10)

customerid  systemloanid  loannumber  \
0  8a2a81a74ce8c05d014cfb32a0da1049     301994762          12   
1  8a85886e54beabf90154c0a29ae757c0     301965204           2   
2  8a8588f35438fe12015444567666018e     301966580           7   
3  8a85890754145ace015429211b513e16     301999343           3   
4  8a858970548359cc0154883481981866     301962360           9   
5  8a8589f35451855401546b0738c42524     301986516           8   
6  8a858e095c59b91b015c5e5cea3719bc     301972027           3   
7  8a858e1158dc4d830158f7bde4f47ea7     301994428          10   
8  8a858e185b4923b4015b4ae48d28646a     301996687           4   
9  8a858e1d5cd58f9e015ceda4bdb63673     301981931           2   

                 approveddate                creationdate  loanamount  \
0  2017-07-25 08:22:56.000000  2017-07-25 07:22:47.000000     30000.0   
1  2017-07-05 17:04:41.000000  2017-07-05 16:04:18.000000     15000.0   
2  2017-07-06 14:52:57.000000  2017-07-06 13:52:51.000000     20000.0   
3  2017-07-27 19:00:41.000000  2017-07-27 18:00:35.000000     10000.0   
4  2017-07-03 23:42:45.000000  2017-07-03 22:42:39.000000     40000.0   
5  2017-07-19 21:46:24.000000  2017-07-19 20:46:18.000000     30000.0   
6  2017-07-10 19:25:33.000000  2017-07-10 18:25:28.000000     10000.0   
7  2017-07-24 23:42:27.000000  2017-07-24 22:42:21.000000     30000.0   
8  2017-07-26 10:53:05.000000  2017-07-26 09:52:57.000000     10000.0   
9  2017-07-17 13:48:42.000000  2017-07-17 12:48:35.000000     10000.0   

   totaldue  termdays referredby good_bad_flag  
0   34500.0        30        NaN          Good  
1   17250.0        30        NaN          Good  
2   22250.0        15        NaN          Good  
3   11500.0        15        NaN          Good  
4   44000.0        30        NaN          Good  
5   39000.0        60        NaN          Good  
6   13000.0        30        NaN          Good  
7   34500.0        30        NaN          Good  
8   11500.0        15        NaN          Good  
9   11500.0        15        NaN          Good

## Viewing data types in dataframes

In [13]:
df_train_prev_loans.dtypes

customerid          object
systemloanid         int64
loannumber           int64
approveddate        object
creationdate        object
loanamount         float64
totaldue           float64
termdays             int64
closeddate          object
referredby          object
firstduedate        object
firstrepaiddate     object
dtype: object

In [12]:
df_train_perf.dtypes

customerid        object
systemloanid       int64
loannumber         int64
approveddate      object
creationdate      object
loanamount       float64
totaldue         float64
termdays           int64
referredby        object
good_bad_flag     object
dtype: object

In [14]:
df_train_demographics.dtypes

customerid                     object
birthdate                      object
bank_account_type              object
longitude_gps                 float64
latitude_gps                  float64
bank_name_clients              object
bank_branch_clients            object
employment_status_clients      object
level_of_education_clients     object
dtype: object

## Obtaining value counts per field

In [15]:
df_train_demographics.count()

customerid                    4346
birthdate                     4346
bank_account_type             4346
longitude_gps                 4346
latitude_gps                  4346
bank_name_clients             4346
bank_branch_clients             51
employment_status_clients     3698
level_of_education_clients     587
dtype: int64

In [16]:
df_train_perf.count()

customerid       4368
systemloanid     4368
loannumber       4368
approveddate     4368
creationdate     4368
loanamount       4368
totaldue         4368
termdays         4368
referredby        587
good_bad_flag    4368
dtype: int64

In [18]:
df_train_prev_loans.count()

customerid         18183
systemloanid       18183
loannumber         18183
approveddate       18183
creationdate       18183
loanamount         18183
totaldue           18183
termdays           18183
closeddate         18183
referredby          1026
firstduedate       18183
firstrepaiddate    18183
dtype: int64

In [21]:
df_train_prev_loans['customerid'].nunique()

4359

In [23]:
df_train_prev_loans['referredby'].nunique()

521

In [24]:
df_train_prev_loans['referredby'].unique()

array([nan, '8a858e6c5c88d145015c89430ef61a5d',
       '8a858ff85bd93919015bda56605652bd',
       '8a858edc59ee87640159eeb9774f1aa7',
       '8a858fc55b2548dd015b286e452c678c',
       '8a858eba5c884d2a015c8bea59385157',
       '8a858fba5cd5feee015cdf5025e10f56',
       '8a858e185b4923b4015b536354895cae',
       '8a858fd85b685607015b7ba5638674c0',
       '8a858f9f5bd99987015be3f53cb442d9',
       '8a858fee5800e000015805aae2fe4ef8',
       '8a858fdc57ab280a0157b277e9ff6d53',
       '8a858e645755e62201575e72c8b77fd5',
       '8a858edb5ad79cc6015ad7c9b29415c8',
       '8a858fc259da7a360159da9011bd0e86',
       '8a858f975c4582c4015c54619e9b7e2b',
       '8a858e6f5bce1023015bce19dbbf038d',
       '8a858f1b5b9136f7015b9f66e5c5327c',
       '8a858e7855113fb701551fe018ce6913',
       '8a858fa3551e78cb01552596fed60cec',
       '8a858f9d5668e3f10156767e9a0f0188',
       '8a858fad5bf85d5e015bf8efc79e33dd',
       '8a858eda5c8863ff015c8b96e5fd74d0',
       '8a858ed75732b681015744c9dee21aae',
      

For 4368 customers, only 4359 have previous loans. The seven without any previous loans will be given zero values for previous loans fields.